pyscf_ipu/experimental/nuclear_gradients.py

```diff
@@ -0,0 +1,92 @@
+) -> Float3:
+    """Generic gradient of a one-electron integral with respect the atom_index center"""
+    axes = jnp.arange(3)
+    lhs_p1 = vmap(a.offset_lmn, (0, None))(axes, 1)
+    t1 = 2 * a.alpha * vmap(primitive_op, (0, None))(lhs_p1, b)
+
+    lhs_m1 = vmap(a.offset_lmn, (0, None))(axes, -1)
+    t2 = jnp.where(a.lmn > 0, a.lmn, jnp.zeros_like(a.lmn))
+    t2 *= vmap(primitive_op, (0, None))(lhs_m1, b)
+    grad_out = t1 - t2
+    return jnp.where(a.atom_index == atom_index, grad_out, jnp.zeros_like(grad_out))
+
+
+def grad_integrate(b: Basis, primitive_op: Callable) -> Float3xNxN:
+    """Generic gradient of one-electron integrals over the basis set"""
+
+    def take_primitives(indices):
+        p = tree_map(lambda x: jnp.take(x, indices, axis=0), primitives)
+        c = jnp.take(coefficients, indices)
+        return p, c
+
+    primitives, coefficients, orbital_index = batch_orbitals(b.orbitals)
+    ii, jj = jnp.meshgrid(*[jnp.arange(b.num_primitives)] * 2, indexing="ij")
+    lhs, cl = take_primitives(ii.reshape(-1))
+    rhs, cr = take_primitives(jj.reshape(-1))
+
+    op = vmap(primitive_op, (None, 0, 0))
+    op = jit(vmap(op, (0, None, None)))
+    atom_indices = jnp.arange(b.structure.num_atoms)
+    out = op(atom_indices, lhs, rhs)
+    out = jnp.sum(out, axis=0)
+
+    out = cl * cr * out.T
+    out = out.reshape(3, b.num_primitives, b.num_primitives)
+    out = segment_sum(jnp.rollaxis(out, 1), orbital_index)
+    out = segment_sum(jnp.rollaxis(out, -1), orbital_index)
+
+    return jnp.rollaxis(out, -1)
+
+
+def grad_overlap_primitives(i: int, a: Primitive, b: Primitive) -> Float3:
+    return grad_primitive_integral(_overlap_primitives, i, a, b)
+
+
+def grad_kinetic_primitives(i: int, a: Primitive, b: Primitive) -> Float3:
+    return grad_primitive_integral(_kinetic_primitives, i, a, b)
+
+
+def grad_overlap_basis(b: Basis) -> Float3xNxN:
+    return grad_integrate(b, grad_overlap_primitives)
+
+
+def grad_kinetic_basis(b: Basis) -> Float3xNxN:
+    return grad_integrate(b, grad_kinetic_primitives)
+
+    def take_primitives(indices):
+        p = tree_map(lambda x: jnp.take(x, indices, axis=0), primitives)
+        c = jnp.take(coefficients, indices)
+        return p, c
+
+    primitives, coefficients, orbital_index = batch_orbitals(b.orbitals)
+    ii, jj = jnp.meshgrid(*[jnp.arange(b.num_primitives)] * 2, indexing="ij")
+    lhs, cl = take_primitives(ii.reshape(-1))
+    rhs, cr = take_primitives(jj.reshape(-1))
+
+    op = vmap(grad_kinetic_primitives, (None, 0, 0))
+    op = jit(vmap(op, (0, None, None)))
+    atom_indices = jnp.arange(b.structure.num_atoms)
+    out = op(atom_indices, lhs, rhs)
+    out = jnp.sum(out, axis=0)
+
+    out = cl * cr * out.T
+    out = out.reshape(3, b.num_primitives, b.num_primitives)
+    out = segment_sum(jnp.rollaxis(out, 1), orbital_index)
+    out = segment_sum(jnp.rollaxis(out, -1), orbital_index)
+
+    return jnp.rollaxis(out, -1)
```

Comment:
Is this now supplanted by the generic one above?

Reply:
yes sorry forgot to push that commit...